### PR TITLE
OCPCLOUD-2693: Support Capacity Reservations by adding 'CapacityReservationId'

### PR DESCRIPTION
**Pull request by athiruma**
This PR supports the Capacity Reservation by passing the `CapaCityReservationId` to the `ec2. RunInstancesInput`.
CapacityReservationId specifies the target Capacity Reservation into which the instance should be launched

### CODE REVIEW

File: pkg/actuators/machine/instances_test.go [hunk 1]
```diff
@@ -1544,3 +1544,35 @@ func TestGetAvalabilityZoneTypeFromZoneName(t *testing.T) {
 		})
 	}
 }
+
+func TestGetCapacityReservationSpecification(t *testing.T) {
```
Comment (theobarberbany) [inline pkg/actuators/machine/instances_test.go:1548]: Ideally, all new tests would use ginkgo + gomega :)

File: pkg/actuators/machine/instances_test.go [hunk 2]
```diff
@@ -1544,3 +1544,35 @@ func TestGetAvalabilityZoneTypeFromZoneName(t *testing.T) {
...
+		capacityReservationID string
+		expectedRequest       *ec2.CapacityReservationSpecification
+	}{
+		{
+			name:                  "with no CapacityReservationID options specified",
+			capacityReservationID: "",
+			expectedRequest:       nil,
+		},
+		{
+			name:                  "with a valid CapacityReservationID specified",
```
Comment (theobarberbany) [inline pkg/actuators/machine/instances_test.go:1561]: Is there any scenario where we may pass an invalid CapacityReservationID? Do we want some verification or testing around that?

**Comment by theobarberbany** [comment]
Hey @athiruma, overall I think this looks good. Just a couple of comments/questions and I think it's good to go!

File: pkg/actuators/machine/instances_test.go (hunk 2 above)
Comment (athiruma) [reply pkg/actuators/machine/instances_test.go:1561]: All the AWS resources start with their short resource name for example cr-***, so I thought no need for verification.

Comment (JoelSpeed) [reply pkg/actuators/machine/instances_test.go:1561]: You still need to verify the user input, what if they change the value to something that isn't the correct format?

Comment (athiruma) [reply pkg/actuators/machine/instances_test.go:1571]: Done, verify that the id starts with the prefix  `cr-`.

File: pkg/actuators/machine/instances.go [hunk 3]
```diff
@@ -639,3 +645,21 @@ func getInstanceMetadataOptionsRequest(providerConfig *machinev1beta1.AWSMachine
...
 	return imdsOptions
 }
+
+func getCapacityReservationSpecification(capacityReservationID string) (*ec2.CapacityReservationSpecification, error) {
+	if capacityReservationID == "" {
+		//  Not targeting any specific Capacity Reservation
+		return nil, nil
+	}
+
+	if !strings.HasPrefix(capacityReservationID, "cr-") {
```
Comment (JoelSpeed) [inline pkg/actuators/machine/instances.go:655]: Are there any other validations we can apply, maximum length? Valid characters? Is there a regex we could use to validate the user input here?

Comment (athiruma) [reply pkg/actuators/machine/instances.go]: right resource-id will be 17char long excluding `cr-` and we check regex with a mix of `lower alpha` and `digits`.

File: pkg/actuators/machine/instances.go [hunk 4]
```diff
@@ -639,3 +646,24 @@ func getInstanceMetadataOptionsRequest(providerConfig *machinev1beta1.AWSMachine
...
+		//  Not targeting any specific Capacity Reservation
+		return nil, nil
+	}
+
+	// Starts with cr-xxxxxxxxxxxxxxxxx with length of 17 characters excluding cr-
+	re := regexp.MustCompile(`^cr-[0-9a-f]{17}$`)
+
+	if !re.MatchString(capacityReservationID) {
+		// It must starts with cr-xxxxxxxxxxxxxxxxx with length of 17 characters excluding cr-
+		return nil, mapierrors.InvalidMachineConfiguration("Invalid value for capacityReservationId: %q, it must starts with cr-", capacityReservationID)
```
Comment (JoelSpeed) [inline pkg/actuators/machine/instances.go:661]:
```suggestion
		return nil, mapierrors.InvalidMachineConfiguration("Invalid value for capacityReservationId: %q, it must starts with cr- followed by 17 lowercase hexadecimal characters", capacityReservationID)
```

File: pkg/actuators/machine/instances_test.go [hunk 5]
```diff
@@ -1544,3 +1545,49 @@ func TestGetAvalabilityZoneTypeFromZoneName(t *testing.T) {
 		})
 	}
 }
+
+func TestGetCapacityReservationSpecification(t *testing.T) {
+	mockCapacityReservationID := "cr-1234a6789d234f6f4"
+	invalidMockCapacityReservationID := "cr-1234"
```
Comment (JoelSpeed) [inline pkg/actuators/machine/instances_test.go:1551]: Can you try a few variations, different lengths, ones with capitals, ones with non hex digits

**Comment by JoelSpeed** [comment]
/approve

Can you please add the relevant Jira card to the PR title please

@theobarberbany can you LGTM if you're ok with this as it is now?

**Comment by JoelSpeed** [comment]
@athiruma You don't have any cards? What's tracking this work? There must be an epic or something so that we can co-ordinate between QE/Docs/PX about the feature. If you don't have anything, please go and make it. Who's the PM on this feature?

**Comment by athiruma** [comment]
> @athiruma You don't have any cards? What's tracking this work? There must be an epic or something so that we can co-ordinate between QE/Docs/PX about the feature. If you don't have anything, please go and make it. Who's the PM on this feature?

yup added to the title.
Thanks @subin for creating a Jira card.

**Pull request merged**